rfc/rfc-91/rfc-91.md

```diff
@@ -0,0 +1,145 @@
+
+JIRA: [HUDI-9122](https://issues.apache.org/jira/browse/HUDI-9122)
+
+## Abstract
+
+Currently in Hudi, distributed locking relies on external systems like Zookeeper, which add complexity and extra dependencies. This RFC introduces a storage-based implementation of the `LockProvider` interface that utilizes conditional writes in cloud storage platforms (such as GCS and AWS S3) to implement a native distributed locking mechanism for Hudi. By directly integrating lock management with cloud storage, this solution reduces operational overhead, and ensures robust coordination during concurrent writes.
+
+## Background
+
+AWS S3 recently introduced conditional writes, and GCS and Azure storage already support them. This RFC leverages these features to implement a distributed lock provider for Hudi using a leader election algorithm. In this approach, each process attempts an atomic conditional write to a file calculated using the table base path. The first process to succeed is elected leader and takes charge of exclusive operations. This method provides a straightforward, reliable locking mechanism without the need for external lock providers.
+
+## Implementation
+
+This design implements a leader election algorithm for Apache Hudi using a single lock file per table stored in cloud storage. Each table’s lock is represented by a JSON file with the following fields:
+- owner: A unique UUID identifying the lock provider instance.
+- expiration: A UTC timestamp indicating when the lock expires.
+- expired: A boolean flag marking the lock as released.
+
+Example lock file path: `s3://bucket/locks/<cleaned-table-base-path>-<short-sha>.json`
+
+Each `LockProvider` must implement `tryLock()` and `unlock()` however we also need to do our own lock renewal, therefore this implementation also has `renewLock()`. The implementation will import a service using reflection which writes to S3/GCS/Azure based on the provided location to write the locks. This ensures the main logic for conditional writes is shared regardless of the underlying storage.
+
+`tryLock()`
+- No Existing Lock: If the lock file doesn’t exist, a new lock file is created with the current instance’s details using a conditional write that only succeeds if the file is absent.
+- Existing Lock – Not Expired: If a valid (non-expired) lock exists, the process refrains from taking the lock.
+- Existing Lock – Expired: If the lock file exists but is expired, a new lock file is conditionally written. This write uses a precondition based on the current file’s unique tag from cloud storage to ensure that no other process has updated it in the meantime.
+
+`renewLock()`
+- Purpose: Periodically extend the lock’s expiration (the heartbeat).
+- Mechanism: Update the lock file’s expiration using a conditional write that verifies the unique tag from the current lock state. If the tag does not match, the renewal fails, indicating that the lock has been lost.
+
+`unlock()`
+- Purpose: Safely release the lock.
```

Comment:
`unlock` would not delete lock file? so once the lock file is created in the first time lock, it would be always there, and will just be updated with conditional write?

Reply:
That's correct. Deleting the lock file is an unnecessary operation, we can just constantly update it.